### PR TITLE
x86mmu: kernel: Validate existing APIs

### PR DESCRIPTION
**Pull request by akhileshupadhyay**
Testcase developed x86 specific to validate API for userbuffer.

Signed-off-by: Akhilesh Kumar Upadhyay <akhilesh.kumarx.upadhyay@intel.com>

### CODE REVIEW

File: tests/kernel/userbufer_validate/src/main.c
Comment (tbursztyka) [inline tests/kernel/userbufer_validate/src/main.c]: 1 tab only for indentation on all these lines.

File: tests/kernel/userbufer_validate/src/userbuffer_validate.c
Comment (tbursztyka) [inline tests/kernel/userbufer_validate/src/userbuffer_validate.c]: parameters should be aligned to last '(' + 1 space, so here the 10, ... should be right belo (void *...

and the comment /* should fail */, put it on its own line.

Apply that rule everywhere

File: tests/kernel/userbufer_validate/src/userbuffer_validate.c
Comment (tbursztyka) [inline tests/kernel/userbufer_validate/src/userbuffer_validate.c]: always break a line before creating a new function. Apply that everywhere

**Review by tbursztyka — CHANGES_REQUESTED** [review]

**Review by pfalcon — CHANGES_REQUESTED** [review]
This patch is named as if it introduces "API to validate user buffer". But it does not, it just adds a test for an existing API. Please make sure that caption corresponds to the content.

File: tests/kernel/userbufer_validate/src/main.c
Comment (akhileshupadhyay) [reply tests/kernel/userbufer_validate/src/main.c]: 1 tab indentation has been corrected.

File: tests/kernel/userbufer_validate/src/userbuffer_validate.c
Comment (akhileshupadhyay) [reply tests/kernel/userbufer_validate/src/userbuffer_validate.c]: breaking rule applied and corrected everywhere.

File: tests/kernel/userbufer_validate/src/userbuffer_validate.c
Comment (akhileshupadhyay) [reply tests/kernel/userbufer_validate/src/userbuffer_validate.c]: 1 line break applied for all functions.

File: tests/kernel/userbufer_validate/src/userbuffer_validate.c
Comment (tbursztyka) [inline tests/kernel/userbufer_validate/src/userbuffer_validate.c]: missed this line.

**Review by tbursztyka — DISMISSED** [review]

File: tests/kernel/userbufer_validate/src/userbuffer_validate.c
Comment (akhileshupadhyay) [reply tests/kernel/userbufer_validate/src/userbuffer_validate.c]: the missed line corrected.

**Comment by andyross** [comment]
Can we clean up the commit message?  This isn't x86 specific, it's actually PC-specific and assumes qemu's memory map.  The API being tested names itself something more like "x86 mmu" and not "user buffer".  Please explain exactly what functionality is being tested (it looks like it's simply setting PDE/PTE bits directly and then validating that they're still set?) in the commit message.

Also: wouldn't a more effective/portable/bitrot-resistant way to do this be to set the protection modes, and then try to access the pages in violation of that and check that the resulting exception is delivered correctly?

File: tests/kernel/userbufer_validate/src/userbuffer_validate.c
Comment (andrewboie) [inline tests/kernel/userbufer_validate/src/userbuffer_validate.c]: why are these addresses hard-coded?
you need to programmatically select memory addresses to test.
there are linker variables defined to help you with this. see include/linker/linker-defs.h

**Review by tyrantlizard77 — CHANGES_REQUESTED** [review]
You need to remove *all* the hard-coded memory addresses. This test shall not depend specifically on QEMU's memory map. Use defined linker variables to determine what memory ranges exist for various regions.

**Review by pfalcon — DISMISSED** [review]
Thanks for updating the commit subject. (Didn't look further, just removing my -1.)

File: tests/kernel/userbufer_validate/src/userbuffer_validate.c
Comment (akhileshupadhyay) [reply tests/kernel/userbufer_validate/src/userbuffer_validate.c]: hard-coded addresses have been removed, also the linker variables have been used

File: tests/kernel/userbufer_validate/src/userbuffer_validate.c
Comment (andrewboie) [inline tests/kernel/userbufer_validate/src/userbuffer_validate.c]: Why do we have these functions here?
We already have _x86_mmu_set_flags(). Use that instead.
You can make these functions call _x86_mmu_set_flags() if you want. But you should not be poking the page tables directly in this file.

File: tests/kernel/userbufer_validate/src/userbuffer_validate.c
Comment (andrewboie) [inline tests/kernel/userbufer_validate/src/userbuffer_validate.c]: Don't need this function use _x86_mmu_set_flags()

File: tests/kernel/userbufer_validate/src/userbuffer_validate.c
Comment (andrewboie) [inline tests/kernel/userbufer_validate/src/userbuffer_validate.c]: Don't need this function use _x86_mmu_set_flags()

File: tests/kernel/userbufer_validate/src/userbuffer_validate.c
Comment (andrewboie) [inline tests/kernel/userbufer_validate/src/userbuffer_validate.c]: Please do not use hard-coded constants when we have macros. Replace 0x1 with BUF_WRITEABLE here and elsewhere

**Review by tyrantlizard77 — CHANGES_REQUESTED** [review]

**Comment by akhileshupadhyay** [comment]
removed x86_mmu_set_pte_us_bit,  x86_mmu_set_pte_us_bit, x86_mmu_set_pde_us_bit  functions.
Used  _x86_mmu_set_flags() instead. Also the macro is used for masks and flags.

**Review by AdithyaBaglody — CHANGES_REQUESTED** [review]
Test cases for Validating  multiple pages  are missing. Each of the  pages should be configured with different page table options(i.e RW/US).

**Review by tyrantlizard77 — CHANGES_REQUESTED** [review]
This coverage is insufficient.

Right now, _x86_mmu_buffer_validate() is broken, it actually always returns true (nonzero) no matter what.

You are inconsistently checking if the return value is 0 or -EPERM. Please see PR #1441

**Review by tyrantlizard77 — CHANGES_REQUESTED** [review]
I'm making some changes to this API so hold off on this PR until those land.

**Comment by andrewboie** [comment]
hi, any update on this?
the APIs are now fixed.

**Comment by akhileshupadhyay** [comment]
am working on it, update after fixing errors.

**Comment by akhileshupadhyay** [comment]
@andrewboie. I have made the changes, included test coverage.

**Comment by akhileshupadhyay** [comment]
@tbursztyka. The proper indentation in missed lines are restructured.

**Comment by andrewboie** [comment]
@AdithyaBaglody can you review this?

File: tests/kernel/userbufer_validate/src/main.c
Comment (andrewboie) [inline tests/kernel/userbufer_validate/src/main.c]: why is this in a separate file???

Comment (akhileshupadhyay) [reply tests/kernel/userbufer_validate/src/main.c]: This will make the code more legible and easy. Also some of the other codes follow same.

**Comment by akhileshupadhyay** [comment]
@AdithyaBaglody . Changes related to multi pages. Removed all the hard coded values. Lines are aligned with one line gap after every new function starts.

**Comment by akhileshupadhyay** [comment]
@AdithyaBaglody  I have added more test cases. This will cover multi pages test.

**Review by AdithyaBaglody — APPROVED** [review]

**Comment by akhileshupadhyay** [comment]
@andrewboie could you please review the changes requested by you.

**Review by kumarvikash1 — APPROVED** [review]
Changes looks fine now, i can see "multi page buffer", "code coverage" and line breaking has been addressed.

**Comment by kumarvikash1** [comment]
@andrewboie  @tbursztyka   : Can you please review the changes. This PR is planned for R-1.10.

File: tests/kernel/userbufer_validate/src/userbuffer_validate.c
Comment (tbursztyka) [inline tests/kernel/userbufer_validate/src/userbuffer_validate.c]: break a line after PRESET_... You do it sometime and sometime not. Apply that eveywhere

**Review by tbursztyka — COMMENTED** [review]
tiny style issue. rest is fine to me.

File: tests/kernel/userbufer_validate/src/userbuffer_validate.c
Comment (akhileshupadhyay) [reply tests/kernel/userbufer_validate/src/userbuffer_validate.c]: @tbursztyka . Could you please check the style issue is addressed and fixed

**Review by tyrantlizard77 — APPROVED** [review]

**Pull request merged**
